golang/cosmos/types/kv_entry.go

```diff
@@ -0,0 +1,111 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+var _ json.Marshaler = &KVEntry{}
+var _ json.Unmarshaler = &KVEntry{}
+
+// KVEntry represents a string key / string value pair, where the value may be
+// missing. The semantics of a missing value are up to the user.
+type KVEntry struct {
+	key   string
+	value *string
+}
+
+// NewKVEntry creates a KVEntry with the provided key and value
+func NewKVEntry(key string, value string) KVEntry {
+	return KVEntry{key, &value}
+}
+
+// NewKVEntryWithNoValue creates a KVEntry with the provided key and no value
+func NewKVEntryWithNoValue(key string) KVEntry {
+	return KVEntry{key, nil}
+}
+
+// UnmarshalJSON interprets the data as a [key: string, value?: string | null]
+// JSON array and updated the current KVEntry.
+// The key must be a string, and the value (if present) must be a string or null.
+//
+// Implements json.Unmarshaler
+// Note: unlike other methods, this accepts a pointer to satisfy
+// the Unmarshaler semantics.
+func (entry *KVEntry) UnmarshalJSON(msg []byte) (err error) {
+	var generic []interface{}
+	err = json.Unmarshal(msg, &generic)
+
+	if err != nil {
+		return err
+	}
+
+	length := len(generic)
+
+	if length != 1 && length != 2 {
+		return fmt.Errorf("array length %d is not a valid KVEntry", length)
+	}
+
+	key, ok := generic[0].(string)
+	if !ok {
+		return fmt.Errorf("invalid entry key: %q", generic[0])
```

Comment:
```suggestion
		return fmt.Errorf("KVEntry key must be a string: %q", generic[0])
```
or maybe we should avoid reflecting back possibly large key/value data:
```suggestion
		return fmt.Errorf("KVEntry key must be a string, not %T", generic[0])
```